File: EmmyLuaApi/UIEventListener.lua
---@class UIEventListener : EventTrigger
---@field public aniClickDown LuaFunction
---@field public aniClickUp LuaFunction
---@field public soundClick LuaFunction
---@field public onClickDown LuaFunction
---@field public onClickUp LuaFunction
---@field public onClick LuaFunction
---@field public onEnter LuaFunction
---@field public onExit LuaFunction
---@field public onInitializePotentialDrag LuaFunction
---@field public onBeginDrag LuaFunction
---@field public onDrag LuaFunction
---@field public onEndDrag LuaFunction
---@field public onDrop LuaFunction
---@field public onSelect LuaFunction
---@field public onUpdateSelected LuaFunction
---@field public onDeselect LuaFunction
---@field public onSubmit LuaFunction
---@field public onCancel LuaFunction
---@field public onScroll LuaFunction
---@field public onMove LuaFunction
local UIEventListener = {}
---@public
---@param _bool bool
---@return void
function UIEventListener:SetClickLock(_bool) end
---@public
---@param eventData PointerEventData
---@return void
function UIEventListener:OnPointerDown(eventData) end
---@public
---@param eventData PointerEventData
---@return void
function UIEventListener:OnPointerUp(eventData) end
---@public
---@param eventData PointerEventData
---@return void
function UIEventListener:OnPointerClick(eventData) end
---@public
---@param eventData PointerEventData
---@return void
function UIEventListener:OnPointerEnter(eventData) end
---@public
---@param eventData PointerEventData
---@return void
function UIEventListener:OnPointerExit(eventData) end
---@public
---@param eventData PointerEventData
---@return void
function UIEventListener:OnInitializePotentialDrag(eventData) end
---@public
---@param eventData PointerEventData
---@return void
function UIEventListener:OnBeginDrag(eventData) end
---@public
---@param eventData PointerEventData
---@return void
function UIEventListener:OnDrag(eventData) end
---@public
---@param eventData PointerEventData
---@return void
function UIEventListener:OnEndDrag(eventData) end
---@public
---@param eventData PointerEventData
---@return void
function UIEventListener:OnDrop(eventData) end
---@public
---@param eventData BaseEventData
---@return void
function UIEventListener:OnSelect(eventData) end
---@public
---@param eventData BaseEventData
---@return void
function UIEventListener:OnUpdateSelected(eventData) end
---@public
---@param eventData BaseEventData
---@return void
function UIEventListener:OnDeselect(eventData) end
---@public
---@param eventData BaseEventData
---@return void
function UIEventListener:OnSubmit(eventData) end
---@public
---@param eventData BaseEventData
---@return void
function UIEventListener:OnCancel(eventData) end
---@public
---@param eventData PointerEventData
---@return void
function UIEventListener:OnScroll(eventData) end
---@public
---@param eventData AxisEventData
---@return void
function UIEventListener:OnMove(eventData) end
---@public
---@param go GameObject
---@return UIEventListener
function UIEventListener.Get(go) end

File: Assets/LuaFramework/Scripts/Utility/UIEventListener.cs
using LuaInterface;
using UnityEngine;
using UnityEngine.EventSystems;

public class UIEventListener : EventTrigger {
	private bool isLockClick = false;

	public LuaFunction aniClickDown = null; //动画按下
	public LuaFunction aniClickUp = null; //动画抬起
	public LuaFunction soundClick = null; //点击音效

	public LuaFunction onClickDown = null;
	public LuaFunction onClickUp = null;
	public LuaFunction onClick = null;
	public LuaFunction onEnter = null;
	public LuaFunction onExit = null;
	public LuaFunction onInitializePotentialDrag = null;
	public LuaFunction onBeginDrag = null;
	public LuaFunction onDrag = null;
	public LuaFunction onEndDrag = null;
	public LuaFunction onDrop = null;
	public LuaFunction onSelect = null;
	public LuaFunction onUpdateSelected = null;
	public LuaFunction onDeselect = null;
	public LuaFunction onSubmit = null;
	public LuaFunction onCancel = null;
	public LuaFunction onScroll = null;
	public LuaFunction onMove = null;

	/// <summary>
	/// 设定点击锁定
	/// </summary>
	/// <param name="_bool">true：上锁  false：解锁</param>
	public void SetClickLock (bool _bool) {
		isLockClick = _bool;
	}

	public override void OnPointerDown (PointerEventData eventData) {
		if (isLockClick) return;
		if (onClickDown != null) onClickDown.Call (gameObject);
		if (aniClickDown != null) aniClickDown.Call ();
	}

	public override void OnPointerUp (PointerEventData eventData) {
		if (isLockClick) return;
		if (onClickUp != null) onClickUp.Call (gameObject);
		if (aniClickUp != null) aniClickUp.Call ();
	}

	public override void OnPointerClick (PointerEventData eventData) {
		if (isLockClick) return;
		if (onClick != null) onClick.Call (gameObject);
		if (soundClick != null) soundClick.Call ();
	}

	public override void OnPointerEnter (PointerEventData eventData) {
		if (onEnter != null) onEnter.Call (gameObject);
	}

	public override void OnPointerExit (PointerEventData eventData) {
		if (onExit != null) onExit.Call (gameObject);
	}

	public override void OnInitializePotentialDrag (PointerEventData eventData) {
		if (onInitializePotentialDrag != null) onInitializePotentialDrag.Call (gameObject);
	}

	public override void OnBeginDrag (PointerEventData eventData) {
		if (onBeginDrag != null) onBeginDrag.Call (gameObject, eventData);
	}

	public override void OnDrag (PointerEventData eventData) {
		if (onDrag != null) onDrag.Call (gameObject, eventData.position, eventData);
	}

	public override void OnEndDrag (PointerEventData eventData) {
		if (onEndDrag != null) onEndDrag.Call (gameObject, eventData);
	}

	public override void OnDrop (PointerEventData eventData) {
		if (onDrop != null) onDrop.Call (gameObject, eventData);
	}

	public override void OnSelect (BaseEventData eventData) {
		if (onSelect != null) onSelect.Call (gameObject);
	}

	public override void OnUpdateSelected (BaseEventData eventData) {
		if (onUpdateSelected != null) onUpdateSelected.Call (gameObject);
	}

	public override void OnDeselect (BaseEventData eventData) {
		if (onDeselect != null) onDeselect.Call (gameObject);
	}

	public override void OnSubmit (BaseEventData eventData) {
		if (onSubmit != null) onSubmit.Call (gameObject);
	}

	public override void OnCancel (BaseEventData eventData) {
		if (onCancel != null) onCancel.Call (gameObject);
	}

	public override void OnScroll (PointerEventData eventData) {
		if (onScroll != null) onScroll.Call (gameObject);
	}

	public override void OnMove (AxisEventData eventData) {
		if (onMove != null) onMove.Call (gameObject, eventData.moveVector);
	}

	public static UIEventListener Get (GameObject go) {
		UIEventListener listener = go.GetComponent<UIEventListener> ();
		if (listener == null) listener = go.AddComponent<UIEventListener> ();
		return listener;
	}
}